### PR TITLE
feat: manual compaction

### LINKED ISSUE #3882: Manually trigger major compaction
### What problem does the new feature solve?

Currently GrepgtimeDB supports triggering compaction using `SELECT COMPACT_TABLE`, but this statement only triggers regular compaction that assigns SSTs into windows and compact those files inside same window together to alleviate the overhead of compactions. We also need to support more compaction strategies when it's manually triggered so that we can enforce a more efficient data layout.

### What does the feature do?

Adds options to `SELECT COMPACT_TABLE` statements so that we can trigger major compactions using strategies like:
- merge all SST files in some time range into one single file
- split SST files strictly by time windows
  - the time window can be inferred from level 0 files or manually specified

We can start with the easy one which splits SST files according to time windows
The syntax would be like:
```sql
SELECT COMPACT_TABLE(some_table, type="strict_window");
```

Before the compaction, suppose the data layout is:
image

We can find that these 3 files overlaps with each other.

And we expect the compaction output is like (given time window is 50):

image

### Implementation challenges

We have to accommodate different compaction startegies in current compaction framework and reuse the scheduling and IO component.

### PR DESCRIPTION
**Pull request by v0y4g3r**
I hereby agree to the terms of the [GreptimeDB CLA](https://github.com/GreptimeTeam/.github/blob/main/CLA.md).

## Refer to a related PR or issue link (optional)
- closes #3882

## What's changed and what's your intention?
This PR adds support for manually trigger different types of compactions.

The syntax is:
```sql

SELECT COMPACT_TABLE("<table_name>", ["<compact_type>"], ["<options>"])
```

Currently, `compact_type` supports `regular` and `strict_window`.
- `regular`: regular compaction like those triggered by flush.
- `strict_window`: split SST files strictly by time window.

The `<options>` argument is type-sepecific compaction options. For `strict_window`, the option value is the compaction window in seconds.

## Checklist

- [x] I have written the necessary rustdoc comments.
- [x] I have added the necessary unit tests and integration tests.
- [x] This PR requires documentation updates.

### CODE REVIEW

File: src/common/function/src/table/flush_compact_table.rs [hunk 1]
```diff
@@ -34,78 +36,147 @@ use crate::ensure_greptime;
...
-            Ok(Value::from(affected_rows as u64))
+    let ValueRef::String(table_name) = params[0] else {
+        return UnsupportedInputDataTypeSnafu {
+            function: "flush_table",
+            datatypes: params.iter().map(|v| v.data_type()).collect::<Vec<_>>(),
         }
+        .fail();
     };
+
+    let (catalog_name, schema_name, table_name) = table_name_to_full_name(table_name, query_ctx)
```
Comment (sunng87) [inline src/common/function/src/table/flush_compact_table.rs:68]: make sure if the table_name is a full-qualified one, it shares same catalog name with query_ctx

Comment (v0y4g3r) [reply src/common/function/src/table/flush_compact_table.rs:71]: > make sure if the table_name is a full-qualified one, it shares same catalog name with query_ctx

Added catalog validation in this commit e48d09780.

File: src/mito2/src/compaction/window.rs [hunk 2]
```diff
@@ -0,0 +1,380 @@
...
+}
+
+/// Calculates timestamp span between start and end timestamp.
+fn file_time_bucket_span(start_sec: i64, end_sec: i64, bucket_sec: i64) -> Vec<i64> {
+    assert!(start_sec <= end_sec);
+
+    // if timestamp is between `[i64::MIN, i64::MIN.align_by_bucket(bucket)]`, which cannot
+    // be aligned to a valid i64 bound, simply return `i64::MIN` rather than just underflow.
+    let mut start_aligned = start_sec.align_by_bucket(bucket_sec).unwrap_or(i64::MIN);
+    let end_aligned = end_sec.align_by_bucket(bucket_sec).unwrap_or(i64::MIN);
```
Comment (evenyag) [inline src/mito2/src/compaction/window.rs:215]: Should we handle overflow by using falling back to `i64::MIN` and `i64::MAX`? They are not aligned by the `bucket_sec` so we might get an incorrect window if we add `bucket_sec` to `start_aligned`.

Maybe we could ignore this file?

File: src/session/src/table_name.rs [hunk 3]
```diff
@@ -12,20 +12,30 @@
...
 pub fn table_name_to_full_name(
     name: &str,
     query_ctx: &QueryContextRef,
 ) -> Result<(String, String, String)> {
     let obj_name = ParserContext::parse_table_name(name, query_ctx.sql_dialect())?;
 
-    table_idents_to_full_name(&obj_name, query_ctx)
+    let (catalog, schema, table) = table_idents_to_full_name(&obj_name, query_ctx)?;
+    ensure!(
+        catalog == query_ctx.current_catalog(),
```
[the remark this answers is not in the document]
Comment (sunng87) [reply src/session/src/table_name.rs:32]: on cloud we now allow cross-schema query/operations. at the moment, schema is just a namespace of tables.

but when we included RBAC, we will need further check for this.

**Review by sunng87 — APPROVED** [review]
LGTM from the permission perspective

File: src/mito2/src/compaction.rs [hunk 4]
```diff
@@ -210,8 +235,22 @@ impl CompactionScheduler {
     /// Schedules a compaction request.
     ///
     /// If the region has nothing to compact, it removes the region from the status map.
-    fn schedule_compaction_request(&mut self, request: CompactionRequest) -> Result<()> {
-        let picker = compaction_options_to_picker(&request.current_version.options.compaction);
+    fn schedule_compaction_request(
+        &mut self,
+        request: CompactionRequest,
+        options: compact_request::Options,
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/compaction.rs:241]: > Could we hold options in the `CompactionRequest`?

It was delibrately removed in [this](https://github.com/v0y4g3r/greptimedb/commit/4080262f56641878e0c0437bbfc874efd97af7e6#diff-f413376968d6fea8034a8259e2f933c1cdfc370f30fa0d49df8a63156bb9cd9bL164) commit

File: src/session/src/table_name.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/session/src/table_name.rs:32]: > on cloud we now allow cross-schema query/operations. at the moment, schema is just a namespace of tables.
> 
> but when we included RBAC, we will need further check for this.

Maybe I should leave a todo here.

File: src/mito2/src/compaction/window.rs (hunk 2 above)
Comment (v0y4g3r) [reply src/mito2/src/compaction/window.rs:216]: > Should we handle overflow by using falling back to `i64::MIN` and `i64::MAX`? They are not aligned by the `bucket_sec` so we might get an incorrect window if we add `bucket_sec` to `start_aligned`.
> 
> Maybe we could ignore this file?

`align_by_bucket` will only underflow for those timestamps near `i64::MIN`, so we can simply assign them to the special `i64::MIN` window.

File: src/mito2/src/compaction.rs [hunk 5]
```diff
@@ -12,37 +12,56 @@
...
 use crate::region::version::{VersionControlRef, VersionRef};
 use crate::region::ManifestContextRef;
 use crate::request::{OptionOutputTx, OutputTx, WorkerRequest};
 use crate::schedule::scheduler::SchedulerRef;
+use crate::sst::file::{FileHandle, FileId, Level};
 use crate::sst::file_purger::FilePurgerRef;
+use crate::sst::version::LevelMeta;
 use crate::worker::WorkerListener;
 
+mod buckets;
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/compaction.rs]: > Put mods before `use`?

fixed in 1f6021037e6cdf18c3981008b6281096e7fcf304

File: src/mito2/src/compaction/window.rs [hunk 6]
```diff
@@ -0,0 +1,380 @@
...
+    let mut buckets = BTreeMap::new();
+
+    for file in files {
+        if file.compacting() {
+            continue;
+        }
+        let (start, end) = file.time_range();
+        let bounds = file_time_bucket_span(
+            start.convert_to(TimeUnit::Second).unwrap().value(),
+            end.convert_to(TimeUnit::Second).unwrap().value(),
```
[the remark this answers is not in the document]
Comment (v0y4g3r) [reply src/mito2/src/compaction/window.rs:196]: fixed in 01830b5b742368b8bb9455285ac4fe864d22a980

**Review by evenyag — APPROVED** [review]

**Comment by sunng87** [comment]
By the way, I remember we have an http api the flush tables, is it removed?

**Pull request merged**